Instant-Messenger/src/server/ConnectionMonitor.cpp

```diff
@@ -0,0 +1,59 @@
+#include "../../include/server/ConnectionMonitor.hpp"
+#include "../../include/util/definitions.hpp"
+
+#ifdef _WIN32
+#include <Windows.h>
+#else
+#include <unistd.h>
+#include <iostream>
+
+#endif
+
+using namespace std;
+
+ConnectionMonitor::ConnectionMonitor() {
+    this->socketMapSemaphore = new Semaphore(1);
+}
+
+void ConnectionMonitor::monitor(int *socket) {
+    std::pair<int *, ConnectionMonitor *> *args = (std::pair<int *, ConnectionMonitor *> *) calloc(1, sizeof(std::pair<int *, ConnectionMonitor *>));
+    args->first = socket;
+    args->second = this;
+
+    keepsMonitoringConnection(args);
+
+    this->socketMapSemaphore->wait();
+    this->socketCountdown.erase(*socket);
+    this->socketMapSemaphore->post();
+}
+
+void * ConnectionMonitor::keepsMonitoringConnection(void *args) {
+    std::pair<int *, ConnectionMonitor *> *pair = (std::pair<int *, ConnectionMonitor *> *) args;
+    int socket = *(int *) pair->first;
+    ConnectionMonitor *_this = (ConnectionMonitor *) pair->second;
+
+    _this->socketMapSemaphore->wait();
+    _this->socketCountdown[socket] = CONNECTION_LIVE;
+    _this->socketMapSemaphore->post();
+
+    bool connectionIsValid;
+    do {
+        sleep(CONNECTION_COUNTDOWN_STEP);
+        _this->socketMapSemaphore->wait();
+        _this->socketCountdown[socket] -= CONNECTION_COUNTDOWN_STEP;
+        _this->socketMapSemaphore->post();
+        connectionIsValid = _this->socketCountdown[socket] > 0;
```

Comment:
Agora entendi o `CONNECTION_COUNTDOWN_STEP`

Não sei, vê tua opinião cmg: não seria mais simples apenas ir testando timestamps? 
i.e. `sleep(KEEP_ALIVE_INTERVAL)` e testar o `timestamp > KEEP_ALIVE_TIMEOUT`

Reply:
Pode ser mais simples. Não me perguntei como fazer a comparação com times pq tinha essa implementação aí na cabeça. Se for algo que dê pra trocar nos próximos 30 min eu o farei.

Reply:
no final vamos estar comparando se a difereça entre dois inteiros são maior do que o timeout configurado. Vamos estar comparando dois inteiros maiores. 
Em verdade eu vos digo dasduas. Tem que usar uma função 'difftime(now,mktime(&old_date));' esse resultado vai ser em segundos.

Reply:
pensando mais um pouquinho, a possibilidade de remover aquele decremento usando comparação de timestamps pode justificar, to vendo aqui

Reply:
feito